File: Binary_Search/LC_700-Search_in_a_Binary_Search_Tree/LC_700-Search_in_a_Binary_Search_Tree.cpp
// #include <bits/stdc++.h>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    // Iterative approach
    // T: O(log n) | O(h), M: O(1)
    // Where n is num of nodes, and h is height of resonably balanced tree
    TreeNode *searchBST(TreeNode *root, int val) {
        while (root) {
            if (root->val < val) 
                root = root->right;     // increase curr node value
            else if (root->val > val)
                root = root->left;      // decrease curr node value
            else                    // Found value
                return root;
        }
        return nullptr;
    }

    /*
    // Recursive approach
    // T: O(log n) | O(h), M: O(1)
    // Where n is num of nodes, and h is height of resonably balanced tree
    TreeNode *searchBST(TreeNode *root, int val) {
        if (!root)
            return nullptr;
        
        if (root->val < val) 
            return searchBST(root->right, val); // increase curr node value
        else if (root->val > val) 
            return searchBST(root->left, val);  // decrease curr node value
        
        return root;    // Found value
    }
    */
};

int main() {
    return 0;
}

File: Binary_Search/LC_700-Search_in_a_Binary_Search_Tree/LC_700-Search_in_a_Binary_Search_Tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # Iterative approach
    # T: O(log n) | O(h), M: O(1)
    # Where n is num of nodes, and h is height of resonably balanced tree
    def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        while root:
            if val > root.val:
                root = root.right   # increase curr node value
            elif val < root.val:
                root = root.left    # decrease curr node value
            else:                # Found value
                return root
        return None
    
    '''
    # Recursive approach
    # T & M: O(log n) | O(h)
    # Where n is num of nodes, and h is height of resonably balanced tree
    def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if not root:
            return None
        
        # Search for value
        if root.val > val:
            return self.searchBST(root.left, val)   # decrease curr val
        elif root.val < val:
            return self.searchBST(root.right, val)  # increase curr val

        return root # Found value
    '''